### PR TITLE
[FIX] removes memory leak issue by removing c code and using numpy/scipy functions

### LINKED ISSUE #6: gradunwarp memory leak
Not the right repo, but issues at https://github.com/Washington-University/gradunwarp are closed. 
(Is there any reason for that? Issues could help communicate on bugs and encourage community to contribute). Is there a more active fork where these should be opened? I thought https://github.com/Washington-University/gradunwarp was the most supported one as it is the version included in an active BIDS app. )

gradunwarp seems to have important memory leaks that might not be detected when running on the command line as python process do not live longer.
The problem seems to originate from the C extension, but this is beyond my expertise.

Thanks.

### PR DESCRIPTION
**Pull request by bpinsard**
Fixes #6 
The "optimized" c-code was not performing any faster than numpy/scipy functions.

TODO:

- [ ] write tests (including testing that these changes do not affect results). 
- [ ] add style/linting actions
- [ ] add actions for running tests automatically
- [ ] add actions to deploy to pypi

### CODE REVIEW

**Comment by glasserm** [comment]
Is this ready for consideration?  Has the patch been tested?

**Comment by bpinsard** [comment]
This is still a draft/WIP, I still need to write tests (including regression tests to assess that the results are consistent with what was previously generated).
I opened the PR to start discussions.

**Comment by coalsont** [comment]
This has merge conflicts, likely due to all the style changes.  It looks like nothing similar has been integrated via another PR, can you tidy this PR by not including the style changes?

**Comment by bpinsard** [comment]
I rebased removing the commit with aggressive black formatting, there are still empty spaces deletions that my editor automatically performed, but that should me more readable.
I had started to work on "regression" tests to see if scipy code was giving similar results as the custom C code, it is not exactly the same results but close enough. This required a dummy spherical harmonics coeffs file, which I tweaked from a real one that cannot be shared due to Siemens copyrights.

File: gradunwarp/core/unwarp_resample.py [hunk 1]
```diff
@@ -314,10 +308,11 @@ def non_linear_unwarp_siemens(self, volshape, dv, dxyz, m_rcs2lai, m_rcs2lai_noh
             out[..., s] = im2
 
         print()
-       
-        img=nib.Nifti1Image(fullWarp,self.m_rcs2ras)
+
+        img=nib.Nifti1Image(fullWarp, self.m_rcs2ras)
         nib.save(img,"fullWarp_abs.nii.gz")
         # return image and the jacobian
+        del vrcsw, vfsl, vxyzw, vrcs, vxyz, vrcsg, fullWarp
```
Comment (coalsont) [inline gradunwarp/core/unwarp_resample.py]: This looks like it may be deleting local variables immediately before a function return (which should cause all local variables to be deleted anyway, right?).  Why is this necessary?

File: setup.py [hunk 2]
```diff
@@ -1,52 +1,24 @@
...
 
 def configuration(parent_package='', top_path=None):
     from numpy.distutils.misc_util import Configuration
     config = Configuration('',parent_package,top_path)
-    config.add_data_files ( *dats )
     return config
 
 setup(name='gradunwarp',
-      version = '1.2.1+HCP',
+      version = '1.3.0+HCP',
```
Comment (coalsont) [inline setup.py]: Please don't bump the version inside a PR.  Also, setup.py in master has changed significantly due to other maintainer edits, please resolve the conflicts.

File: gradunwarp/core/unwarp_resample.py (hunk 1 above)
Comment (bpinsard) [reply gradunwarp/core/unwarp_resample.py]: Remains of function based memory profiling, indeed unnecessary.

File: gradunwarp/core/tests/data/gradunwarp_coeffs.grad [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#****************************************************************************/
+
+#*] END: */
+dummy, dummy , Gx,y,z = 80/80 mT/m
+win_low = 0, win_high = 0, win_algo = 0, win_dummy = 2;
+0.25 m = R0, lnorm = 4? A(1,0) = B(1,1) = A(1,1) = 0;
+0 = CoSyMode,
+NO.    TYPE                  SPECTRUM                                AXIS
+
+1 A( 3, 0)           -0.0510000                  z
```
Comment (coalsont) [inline gradunwarp/core/tests/data/gradunwarp_coeffs.grad]: Were these all replaced with random values?  Sounds like we should get sign-off from the vendor that you used as a starting point for this dummy file...

File: gradunwarp/core/utils.py [hunk 4]
```diff
@@ -54,6 +54,12 @@ def transform_coordinates(A, M):
     B1, B2, B3 = _transform_coordinates(A1, A2, A3, M)
     return CoordsVector(B1, B2, B3)
 
+from nibabel.affines import apply_affine
+
+def transform_coordinates(A, M):
+    vecs = np.asanyarray([A.x,A.y,A.z]).T
+    vecs_trans = apply_affine(M, vecs)
+    return CoordsVector(vecs_trans[...,0].T, vecs_trans[...,1].T, vecs_trans[...,2].T)
```
Comment (effigies) [inline gradunwarp/core/utils.py]: This can be done more simply:

```suggestion
    transformed = apply_affine(M, np.stack(A).T).T
    return CoordsVector(*transformed)
```

File: gradunwarp/core/unwarp_resample.py [hunk 5]
```diff
@@ -196,7 +189,7 @@ def non_linear_unwarp_siemens(self, volshape, dv, dxyz, m_rcs2lai, m_rcs2lai_noh
         dvz = np.zeros((nr, nc), dtype=np.float32)
         im_ = np.zeros((nr, nc), dtype=np.float32)
         # init jacobian temp image
-        vc, vr = utils.meshgrid(np.arange(nc), np.arange(nr))
+        vc, vr = np.meshgrid(np.arange(int(nc)), np.arange(int(nr)))
```
Comment (effigies) [inline gradunwarp/core/unwarp_resample.py]: These are already ints, since they come from a volume shape.

File: gradunwarp/core/unwarp_resample.py [hunk 6]
```diff
@@ -411,16 +404,16 @@ def siemens_B(alpha, beta, x1, y1, z1, R0):
         f = np.power(r / R0, n)
         for m in range(0, n + 1):
             f2 = alpha[n, m] * np.cos(m * phi) + beta[n, m] * np.sin(m * phi)
-            _ptemp = utils.legendre(n, m, np.cos(theta))
-            #_ptemp = scipy.special.lpmv(m, n, np.cos(theta))
+            #_ptemp = utils.legendre(n, m, np.cos(theta))
+            _p = scipy.special.lpmv(m, n, np.cos(theta))
```
Comment (effigies) [inline gradunwarp/core/unwarp_resample.py]: This `np.cos(theta)` is a constant, and `theta = np.arccos(z1 / r)` above. Possibly out-of-scope with this PR, but would it make sense to set `cosine_theta = z1 / r` above and use that constant here?

File: gradunwarp/core/unwarp_resample.py [hunk 7]
```diff
@@ -411,16 +404,16 @@ def siemens_B(alpha, beta, x1, y1, z1, R0):
...
             normfact = 1
             # this is Siemens normalization
             if m > 0:
                 normfact = math.pow(-1, m) * \
                 math.sqrt(float((2 * n + 1) * factorial(n - m)) \
                           / float(2 * factorial(n + m)))
-            _p = normfact * _ptemp
-            b = b + f * _p * f2
+            _p *= normfact
```
Comment (effigies) [inline gradunwarp/core/unwarp_resample.py]: It seems cleaner to me to move `_p *= normfact` into the `if` block above, rather than unconditionally multiply.

```suggestion
                _p *= normfact
```

File: gradunwarp/core/unwarp_resample.py [hunk 8]
```diff
@@ -86,15 +87,15 @@ def eval_spharm_grid(self, vendor, coeffs):
         g_xyz2rcs = np.linalg.inv(g_rcs2xyz)
 
         # indices into the gradient displacement vol
-        gr, gc, gs = utils.meshgrid(np.arange(numpoints), np.arange(numpoints),
-                                 np.arange(numpoints), dtype=np.float32)
+        gr, gc, gs = np.meshgrid(np.arange(numpoints), np.arange(numpoints),
+                                 np.arange(numpoints))
```
Comment (effigies) [inline gradunwarp/core/unwarp_resample.py]: These are not used in this function. No point creating them.

**Review by effigies — COMMENTED** [review]
Seems like a pretty straightforward improvement. Do you have benchmarks? Or would you just time the regression test before/after the refactor?

@coalsont If this is accepted and extensions no longer need to be compiled, I will open a PR to update the CI to only build a single, pure Python wheel before you cut a new release.

**Comment by coalsont** [comment]
Looks like the test may be looking in the wrong folder for the coefficients file when run as a github action.

**Comment by effigies** [comment]
The test data doesn't get packaged by default. You should be able to add this to `setup.cfg`:

```ini
[options.package_data]
gradunwarp.core.tests = data/*
```

Note that this will increase package size from about 16k to 1.1M. We could split out the `tests/` directory so that the tests and their data are packaged in the source distribution but not installed into the package. I've fiddled around with this and can open a PR against @bpinsard's branch. If that's adding too much complication to this PR, I can wait until it's done and submit another to reduce the package size.

**Comment by bpinsard** [comment]
Thanks for the in depth code review, there are indeed a lot of suboptimal parts in the original code which is itself somewhat hard to read.

> Seems like a pretty straightforward improvement. Do you have benchmarks? Or would you just time the regression test before/after the refactor?

I had timed it interactively and it was equivalent if not faster, but I have no actual values. It seems like the C extension were created a long time ago when scipy/numpy equivalent did not exists. We can expect numpy/scipy code to be of better quality/performance/robustness.
The approach you suggest with timing of the regression test seems the easiest way, I will give it a try.

> @coalsont If this is accepted and extensions no longer need to be compiled, I will open a PR to update the CI to only build a single, pure Python wheel before you cut a new release.

File: setup.py [hunk 9]
```diff
@@ -8,14 +8,5 @@
 
 
 setup(
-    ext_modules=[
```
Comment (effigies) [inline setup.py:11]: Can you remove the `Extension` and `get_include` imports while you're here?

And in case you missed it above, the tests are failing on CI because we test the installed package. You need to add the following to `setup.cfg`:

```ini
[options.package_data]
gradunwarp.core.tests = data/*
```

File: gradunwarp/core/tests/test_regression.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+def test_siemens_B():
+    gradfile = 'gradunwarp/core/tests/data/gradunwarp_coeffs.grad'
+    siemens_coeffs = coeffs.get_coefficients('siemens', gradfile)
+    R0 = siemens_coeffs.R0_m  * 1000
+
+    vec = np.linspace(-300, 300, 60, dtype=np.float32)
+    x, y ,z = np.meshgrid(vec, vec, vec)
+
+    bx = siemens_B(siemens_coeffs.alpha_x, siemens_coeffs.beta_x, x, y, z, R0)
+    ref_bx = np.load('gradunwarp/core/tests/data/siemens_B_output.npz')['bx']
```
Comment (effigies) [inline gradunwarp/core/tests/test_regression.py]: This is going to be brittle, requiring the tests to be run with the working directory set to the repository root. `__file__` is not the cleanest way of referencing packaged files, but it works across all supported versions of Python:

```suggestion

import os

DATA_DIR = os.path.join(os.path.dirname(__file__), 'data')

def test_siemens_B():
    gradfile = os.path.join(DATA_DIR, 'gradunwarp_coeffs.grad')
    siemens_coeffs = coeffs.get_coefficients('siemens', gradfile)
    R0 = siemens_coeffs.R0_m  * 1000

    vec = np.linspace(-300, 300, 60, dtype=np.float32)
    x, y ,z = np.meshgrid(vec, vec, vec)

    bx = siemens_B(siemens_coeffs.alpha_x, siemens_coeffs.beta_x, x, y, z, R0)
    ref_bx = np.load(os.path.join(DATA_DIR, 'siemens_B_output.npz'))['bx']
```

**Review by effigies — COMMENTED** [review]
Some packaging notes.

**Comment by bpinsard** [comment]
If you have more suggestions on making the regression test better, or understanding why using scipy legendre give difference at the 5th decimal, I am all ears. Ideally we should try to decrease that discrepancy, but maybe that's impossible.

**Comment by effigies** [comment]
float32 has 7.22 decimal digits of precision. If your values are >10, the 7th digit will be the fifth digit after the decimal, which is what `assert_array_almost_equal` is actually checking. We know that Krish's legendre used floats; if scipy's legendre uses doubles for some portions of the calculation, that could be a source of difference. Also, fp arithmetic is not associative, so any reordering of operations in the implementation could be to blame.

What if you try `assert_allclose`? That defaults to a relative tolerance of 1e-7, which is more suited to comparing the results of single-precision arithmetic.

**Comment by bpinsard** [comment]
Thanks for your input.

`assert_allclose` gives the following
```
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 148042 / 216000 (68.5%)
E           Max absolute difference: 2.86567956e-05
E           Max relative difference: 0.03226674
E            x: array([[[-4.727971, -7.04162 , -8.64953 , ..., -8.649536, -7.04162 ,
E                    -4.727949],
E                   [-3.05993 , -5.125746, -6.616065, ..., -6.61608 , -5.125739,...
E            y: array([[[-4.727948, -7.041609, -8.649537, ..., -8.649537, -7.041609,
E                    -4.727948],
E                   [-3.05992 , -5.12574 , -6.616078, ..., -6.616078, -5.12574 ,...

```

Just the fact of not inefficiently computing `cos(argcos)` amplified the max discrepancy by one decimal.

I guess it depends on what it is we want to achieve, I would say we want close results, but how close do we want it. 
It is not a true regression test, more like a way to check that this large refactor does not breaks badly the algorithm.

Also that displacement value in mm is to be considered in the context of MRI resolution. When interpolating data using such field it will likely result in very minor differences too. 
Should we choose the threshold based on usual MRI resolution?

**Comment by effigies** [comment]
This is not my project, so I can't say what the threshold should be for an acceptable change. If I were, I think my response would be:

1) If a user wants bit-for-bit reproducibility, they should use the same version of the software. Some change across versions is acceptable.
2) It would be good to understand the changes piecemeal.
   * What is the change from the custom `legendre()` implementation vs `scipy.special.lpmv()`, all else held constant? Here, I suspect that the advantage of using an externally vetted implementation would outweigh a small change in results even if they exceed `np.allclose()`'s default tolerances.
   * For further changes, it would be nice to see they either have no effect on the output or that they can be argued for in terms of *reducing* the accumulation of floating point error. Removing a `cos(arccos(x))` round trip is an example where I would say the error is definitively reduced. But this can only be argued in isolation; alongside a bunch of other changes, it's hard to say.

Given the suggestion that changes in displacements of ~1e-5mm are going to have a negligible impact on the actual interpolation, perhaps an additional regression test would be to correct some image. A 1mm^3 resolution MNI template, for example, would allow you to both visually assess and quantify the change in distortion. If that image is `allclose()`, that would help justify a more lax threshold for these specific outputs.

But all this is the call of the maintainers. @coalsont @glasserm WDYT?

**Comment by coalsont** [comment]
@mharms thoughts on simplifying the gradunwarp code in future releases in a way that slightly changes the produced warpfield displacements by up to 30 nanometers?

I think that should be fine, but I'm more worried that Siemens may not be satisfied with how the dummy test coefficients file was constructed (if it is insufficiently different from the real one used as a starting point, such that it might still reveal things about the construction of their coils).

**Comment by mharms** [comment]
Yes, I agree that the tolerances being discussed here seem fine, and the differences being discussed not relevant from a practical perspective.

I can't comment on the "test" coefficient file, either whether it is "insufficiently different", or conversely whether is it so different from the specification of a real gradient coil that the resulting regression test is actually not meaningful.

**Comment by bpinsard** [comment]
Regarding the coefficient file, I guess we could try to make one from scratch.
Are there any references that could help understand on which base the current software was implemented?

While I was trying to reimplement the coeff parser https://github.com/Washington-University/gradunwarp/blob/master/gradunwarp/core/coeffs.py#L119 to be more robust (using regexp to read lines) I am really puzzled about this lines https://github.com/Washington-University/gradunwarp/blob/master/gradunwarp/core/coeffs.py#L138-L144 that do skip a number of coefficient in the dummy file (which has the same structure to the one from our scanner). 
So, for instance, it will only start taking data at ` 4 A( 9, 0) ` in the dummy file above.
Are there any reason to do so? Did it also skip these lines for the HCP data or did HCP scanner coeff file structure was different from ours. 

Thanks for your insight!

**Comment by bpinsard** [comment]
I wrote a dummier grad file, way less coefficients (low order of spherical harmonics I guess), faster to test as well.
Rewrote the git history to remove the old file to prevent any lawsuit :D .

**Comment by coalsont** [comment]
That does look much less related to an actual scanner, thanks.

The history rewrite appears to have diverged at 1.2.1 and made copies of many master commits, while current master does not appear to have the coefficients file in question (and no other commit in any branch appears to create any `*.grad` file).  Can you try to redo that from current master, rather than 1.2.1?  Maybe the merge commit messed with things...you could cherry-pick just the commits you made...

**Comment by bpinsard** [comment]
rebase done.

**Comment by coalsont** [comment]
Rebase looks proper, code looks good at a glance.  Looks like setup.py is still unhappy, seems to contain remnants of the C code modules.

File: gradunwarp/core/coeffs.py [hunk 11]
```diff
@@ -78,137 +78,74 @@ def get_siemens_coef(cfile):
...
-
-    # just return R0_m but also txt_var_map is returned
+    with open(gfile, 'r') as gf:
+        for line in gf:
+            re_search = re.search("(?P<no>\d+)[\s]+(?P<aorb>[AB])[\s]*\(\s*(?P<x>\d+),\s*(?P<y>\d+)\)\s+(?P<spectrum>[\-]?\d+\.\d+)\s+(?P<axis>[xyz])", line)
+            if re_search:
+                re_res = re_search.groupdict()
+                alphabeta = 'alpha' if re_res['aorb'] == 'A' else 'beta'
+                x, y = int(re_res['x']),int(re_res['y'])
+                txt_var_map[f"{alphabeta}_{re_res['axis']}"][x, y] = float(re_res['spectrum'])
```
Comment (coalsont) [inline gradunwarp/core/coeffs.py]: Looks like python2 doesn't like `f"..."`.  Can this be easily written compatibly?

Comment (bpinsard) [reply gradunwarp/core/coeffs.py]: Oh! I didn't realize that it still aim at supporting python2. Will fix that!

File: gradunwarp/core/unwarp_resample.py [hunk 12]
```diff
@@ -268,36 +245,26 @@ def non_linear_unwarp_siemens(self, volshape, dv, dxyz, m_rcs2lai, m_rcs2lai_noh
...
-                                                  output=im_,
-                                                  order=self.order)
+            ndimage.map_coordinates(self.vol,
+                                    vrcsw,
+                                    output=im_,
+                                    order=self.order)
             # find NaN voxels, and set them to 0
-            im_[np.where(np.isnan(im_))] = 0.
-            im_[np.where(np.isinf(im_))] = 0.
+            np.nan_to_num(im_, copy=False, posinf=0, neginf=0)
```
Comment (coalsont) [inline gradunwarp/core/unwarp_resample.py]: There are python2 tests, so I assume it was working previously.  Looks like the python2 version of numpy doesn't know about `posinf` here?

Comment (effigies) [reply gradunwarp/core/unwarp_resample.py]: Just for reference, the last version of numpy to support Python 2 was [1.16](https://numpy.org/doc/1.16/).

Comment (mharms) [reply gradunwarp/core/unwarp_resample.py]: Maybe we should just drop support of Python 2?

Comment (coalsont) [reply gradunwarp/core/unwarp_resample.py]: It looks like he fixed it, so I don't see a benefit to dropping python2 immediately.

**Comment by coalsont** [comment]
Looks ready to merge, but I'll give it a day for any other opinions.
